File: answer.jl
module Answer
function divisor(number)
    answer = []

    for i = 1:number
        if (number % i) == 0
            push!(answer, i)
        end
    end

    answer
end

function isabundantnumber(number)
    sum(divisor(number)) - number > number
end

function isdeficientnumber(number)
    sum(divisor(number)) - number < number
end

function isprime(number)
    for i = 2:floor(sqrt(number))
        if (number % i) == 0
            return false
        end
    end

    return true
end

function collatz(number)
    result = [number]

    while true
        if ((number % 2) == 0)
            number = number / 2
        else
            number = number * 3 + 1
        end

        push!(result, number)

        if number == 1
            return result
        end
    end
end

function q1(number)
    divisor(number)
end

function q2()
    number = 100000
    while number > 0
        if isabundantnumber(number)
            return number
        end

        number -= 1
    end
end

function q3()
    number = 1

    while true
        if size(collatz(number), 1) == 256
            return number
        end

        number += 1
    end
end


function q4(number)
    isprime(number)
end
end

File: answer.go
package answer

import "os"

func divisor(n int) []int {
	var x []int

	for i := 1; i <= n; i++ {
		if (n % i) == 0 {
			x = append(x, i)
		}
	}

	return x
}

func isAbundant(n int) bool {
	var sum int = 0

	for _, x := range divisor(n) {
		sum += x
	}

	if sum-n > n {
		return true
	} else {
		return false
	}
}

func collatz(n int) []int {
	var x []int = []int{n}

	for {
		if (n % 2) == 0 {
			n = n / 2
		} else {
			n = n*3 + 1
		}

		x = append(x, n)

		if n == 1 {
			return x
		}
	}
}

func q1(n int) []int {
	return divisor(n)
}

func q2(n int) int {
	for {
		if isAbundant(n) {
			return n
		}

		if n < 0 {
			os.Exit(1)
		}

		n--
	}
}

func q3(n int) int {
	for {
		if len(collatz(n)) == 256 {
			return n
		}

		n++
	}
}
